### PR TITLE
[InstCombine] Resolve TODO: Remove one-time check if other logic operand (Y) is constant

### PR DESCRIPTION
**Pull request by AZero13**
By using ```match(W, m_ImmConstant())```, we do not need to worry about one-time use anymore.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
Tests?

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp [hunk 1]
```diff
@@ -368,10 +368,14 @@ static Instruction *foldShiftOfShiftedBinOp(BinaryOperator &I,
...
-  // TODO: Remove the one-use check if the other logic operand (Y) is constant.
   Value *X, *Y;
-  auto matchFirstShift = [&](Value *V) {
+  auto matchFirstShift = [&](Value *V, Value *W) {
     APInt Threshold(Ty->getScalarSizeInBits(), Ty->getScalarSizeInBits());
+    if (isa<Constant>(W)) {
+      return match(V, (m_BinOp(ShiftOpcode, m_Value(X), m_Constant(C0)))) &&
+             match(ConstantExpr::getAdd(C0, C1),
+                   m_SpecificInt_ICMP(ICmpInst::ICMP_ULT, Threshold));
+    }
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:378]: Replace the `m_OneUse` with something like `(isa<Constant>(W) || V->hasOneUse())` instead of duplicating the code. Also better replace `isa<Constant>(W)` with `match(W, m_ImmConstant())`.

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp]: Done!

**Comment by AtariDreams** [comment]
> Tests?

Sure. I added a whole bunch.

File: llvm/test/Transforms/InstCombine/shift-logic.ll [hunk 2]
```diff
@@ -16,6 +16,19 @@ define i8 @shl_and(i8 %x, i8 %y) {
...
+; CHECK-NEXT:    [[SH0:%.*]] = shl i8 [[X:%.*]], [[Y:%.*]]
+; CHECK-NEXT:    [[R:%.*]] = and i8 [[SH0]], [[Y]]
+; CHECK-NEXT:    [[SH1:%.*]] = shl i8 [[R]], 2
+; CHECK-NEXT:    ret i8 [[SH1]]
+;
+  %sh0 = shl i8 %x, %y
+  %r = and i8 %sh0, %y
+  %sh1 = shl i8 %r, 2
+  ret i8 %sh1
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/shift-logic.ll:30]: I don't understand this test coverage. This test (and all the others you added) are not affected by the transform.

This transform always requires the shift amount to be constant. The difference would be in having a constant operand on the `and` here, not the `shl`. We would also need a multi-use scenario (see tests using `@use`).

File: llvm/test/Transforms/InstCombine/shift-logic.ll [hunk 3]
```diff
@@ -16,6 +16,16 @@ define i8 @shl_and(i8 %x, i8 %y) {
...
 
+define i8 @shl_and_constant(i8 %x) {
+; CHECK-LABEL: @shl_and_constant(
+; CHECK-NEXT:    ret i8 0
+;
+  %sh0 = shl i8 %x, 3
+  %r = and i8 %sh0, 2
+  %sh1 = shl i8 %r, 2
+  ret i8 %sh1
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/shift-logic.ll:27]: All of these tests are still irrelevant, because they don't test a multi-use scenario. They will behave the same before and after your patch.

File: llvm/test/Transforms/InstCombine/shift-logic.ll [hunk 4]
```diff
@@ -514,3 +857,53 @@ define <2 x i64> @lshr_sub_undef(<2 x i64> %x, <2 x i64> %py) {
...
+; CHECK-LABEL: @multi_use_constant(
+; CHECK-NEXT:    [[SH0:%.*]] = lshr i8 [[X:%.*]], 3
+; CHECK-NEXT:    [[R:%.*]] = add nuw nsw i8 [[SH0]], -42
+; CHECK-NEXT:    call void @func(i8 [[R]])
+; CHECK-NEXT:    [[SH1:%.*]] = lshr i8 [[R]], 2
+; CHECK-NEXT:    ret i8 [[SH1]]
+;
+  %sh0 = lshr i8 %x, 3
+  %r = sub i8 %sh0, 42
+  call void @func(i8 %r)
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/shift-logic.ll:884]: This is getting closer, but still testing the wrong thing. The transform always requires the binop to be one-use, your change is about the *shift*. You should pass `i8 %sh0` instead of `i8 %r` here.

**Comment by AtariDreams** [comment]
@nikic Alright. I think I got this right. Though I am concerned about the shl undef tests that seem to fold things into poison values. I do not know if this is what we want.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp [hunk 5]
```diff
@@ -368,12 +368,17 @@ static Instruction *foldShiftOfShiftedBinOp(BinaryOperator &I,
...
-    return match(V,
-                 m_OneUse(m_BinOp(ShiftOpcode, m_Value(X), m_Constant(C0)))) &&
+    bool matchFirst;
+    if (match(W, m_ImmConstant()))
+      matchFirst = match(V, (m_BinOp(ShiftOpcode, m_Value(X), m_Constant(C0))));
+    else
+      matchFirst =
+          match(V, m_OneUse(m_BinOp(ShiftOpcode, m_Value(X), m_Constant(C0))));
+
+    return matchFirst &&
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:381]:
```suggestion
    return match(V, m_BinOp(ShiftOpcode, m_Value(X), m_Constant(C0))) &&
           (V->hasOneUse() && match(W, m_ImmConstant()))
```
No need to duplicate the match.

File: llvm/test/Transforms/InstCombine/shift-logic.ll [hunk 6]
```diff
@@ -16,6 +14,23 @@ define i8 @shl_and(i8 %x, i8 %y) {
...
+; CHECK-NEXT:    [[R:%.*]] = shl i8 [[X]], 5
+; CHECK-NEXT:    [[SH1:%.*]] = and i8 [[R]], -96
+; CHECK-NEXT:    ret i8 [[SH1]]
+;
+  %sh0 = shl i8 %x, 3
+  %r = and i8 %sh0, 42
+  call void @use(i8 %sh0)
+  %sh1 = shl i8 %r, 2
+  ret i8 %sh1
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/shift-logic.ll:30]: Okay, this is testing the right thing now. However, we don't need a multi-use variant of every single test.

It would be enough to have just one multi-use test with a constant (folds) and another with a variable (does not fold).

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp [hunk 7]
```diff
@@ -689,7 +690,8 @@ static Value *getShiftedValue(Value *V, unsigned NumBits, bool isLeftShift,
   IC.addToWorklist(I);
 
   switch (I->getOpcode()) {
-  default: llvm_unreachable("Inconsistency with CanEvaluateShifted");
```
Comment (jrtc27) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:692]: Don't reformat unrelated code

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp [hunk 8]
```diff
@@ -727,8 +729,8 @@ static Value *getShiftedValue(Value *V, unsigned NumBits, bool isLeftShift,
     IC.InsertNewInstWith(Neg, I->getIterator());
     unsigned TypeWidth = I->getType()->getScalarSizeInBits();
     APInt Mask = APInt::getLowBitsSet(TypeWidth, TypeWidth - NumBits);
-    auto *And = BinaryOperator::CreateAnd(Neg,
```
Comment (jrtc27) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:730]: Ditto

File: llvm/test/Transforms/InstCombine/shift-logic.ll [hunk 9]
```diff
@@ -1,8 +1,6 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
 ; RUN: opt < %s -passes=instcombine -S | FileCheck %s
 
-declare void @use(i64)
```
Comment (jrtc27) [inline llvm/test/Transforms/InstCombine/shift-logic.ll:4]: Why?

File: llvm/test/Transforms/LoopVectorize/AArch64/sve-widen-phi.ll [hunk 10]
```diff
@@ -27,43 +27,43 @@ define void @widen_ptr_phi_unrolled(ptr noalias nocapture %a, ptr noalias nocapt
 ; CHECK-NEXT:    [[N_VEC:%.*]] = and i64 [[DOTNEG]], [[N]]
 ; CHECK-NEXT:    [[TMP3:%.*]] = shl i64 [[N_VEC]], 3
 ; CHECK-NEXT:    [[IND_END:%.*]] = getelementptr i8, ptr [[C:%.*]], i64 [[TMP3]]
-; CHECK-NEXT:    [[TMP25:%.*]] = call i64 @llvm.vscale.i64()
-; CHECK-NEXT:    [[TMP26:%.*]] = shl nuw nsw i64 [[TMP25]], 3
+; CHECK-NEXT:    [[TMP4:%.*]] = call i64 @llvm.vscale.i64()
```
Comment (jrtc27) [inline llvm/test/Transforms/LoopVectorize/AArch64/sve-widen-phi.ll:30]: At a glance this whole file looks like just renumbering; if the old test still passes, don't touch it?

**Review by jrtc27 — CHANGES_REQUESTED** [review]
You have a commit to precommit tests, but the second commit doesn't touch any of the tests?

Also the precommit commit will need to be its own PR otherwise it'll get squashed when merging.

File: llvm/test/Transforms/LoopVectorize/AArch64/sve-widen-phi.ll (hunk 10 above)
Comment (AtariDreams) [reply llvm/test/Transforms/LoopVectorize/AArch64/sve-widen-phi.ll:30]: So, what happened is that this test failed when I made an earlier edit, so I tried to update the tests. But then when I realized a bug happened, I fixed the bug and regenerated the test to revert back. I can undo this change.

**Comment by jrtc27** [comment]
> > You have a commit to precommit tests, but the second commit doesn't touch any of the tests?
> > Also the precommit commit will need to be its own PR otherwise it'll get squashed when merging
> 
> I thought squashed while merging is what we wanted?

Not when you *want* two commits, like in this case, such that the history clearly shows the test differences.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp (hunk 7 above)
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:692]: I don't know why clang-format does that but I am trying to remove these changes.

Comment (jrtc27) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:692]: Use git clang-format rather than formatting the whole file. Or just `git add -p` / `git checkout -p` / etc to select which diffs you want.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp (hunk 5 above)
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp]: Not done.

File: llvm/test/Transforms/InstCombine/shift-logic.ll (hunk 6 above)
Comment (nikic) [reply llvm/test/Transforms/InstCombine/shift-logic.ll]: Not done.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp (hunk 5 above)
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp]: Your suggestion changes the behavior of an unrelated test and undoes the transform.

File: llvm/test/Transforms/InstCombine/shift-logic.ll (hunk 6 above)
Comment (nikic) [reply llvm/test/Transforms/InstCombine/shift-logic.ll]: (And don't use vector tests for this purpose.)

File: llvm/test/Transforms/InstCombine/shift-logic.ll [hunk 11]
```diff
@@ -29,6 +31,21 @@ define <2 x i8> @shl_and_nonuniform(<2 x i8> %x, <2 x i8> %y) {
...
+; CHECK-NEXT:    [[TMP1:%.*]] = shl <2 x i8> [[X]], <i8 5, i8 4>
+; CHECK-NEXT:    [[SH1:%.*]] = and <2 x i8> [[TMP1]], <i8 -88, i8 42>
+; CHECK-NEXT:    ret <2 x i8> [[SH1]]
+;
+  %sh0 = shl <2 x i8> %x, <i8 3, i8 4>
+  %r = and <2 x i8> %sh0, <i8 42, i8 42> ; constant operand on the 'and'
+  call void @use1(<2 x i8> %sh0)
+  %sh1 = shl <2 x i8> %r, <i8 2, i8 0>
+  ret <2 x i8> %sh1
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/shift-logic.ll:47]: Okay, this is close to what we want now. Can you please move this next to `lshr_or_extra_use` and make it more similar to that test (i.e. use same types/operations, just with the binop operand being a constant)? That way we get the clear symmetry between the transform happening in one case but not the other.

File: llvm/test/Transforms/InstCombine/shift-logic.ll [hunk 12]
```diff
@@ -240,6 +240,21 @@ define i32 @lshr_or_extra_use(i32 %x, i32 %y, ptr %p) {
   ret i32 %sh1
 }
 
+define i32 @lshr_or_extra_use_multiuse(i32 %x, ptr %p) {
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/shift-logic.ll:243]: Ah sorry, I now see that the lshr_or_extra_use test has the multi-use on the binop, not the shift. So please make this `store i32 %sh0, ptr %p` and then you'll have to add a second test that's the same just with the variable `%y` operand in the or.

File: llvm/test/Transforms/InstCombine/shift-logic.ll (hunk 11 above)
Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/shift-logic.ll]: @nikic ok

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/shift-logic.ll]: Done

File: llvm/test/Transforms/InstCombine/shift-logic.ll (hunk 12 above)
Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/shift-logic.ll:243]: Done!

File: llvm/test/Transforms/InstCombine/shift-logic.ll [hunk 13]
```diff
@@ -240,6 +240,35 @@ define i32 @lshr_or_extra_use(i32 %x, i32 %y, ptr %p) {
...
+; CHECK-LABEL: @lshr_or_extra_use_multiuse(
+; CHECK-NEXT:    [[SH0:%.*]] = lshr i32 [[X:%.*]], 5
+; CHECK-NEXT:    store i32 [[SH0]], ptr [[P:%.*]], align 4
+; CHECK-NEXT:    [[SH1:%.*]] = lshr i32 [[X]], 12
+; CHECK-NEXT:    ret i32 [[SH1]]
+;
+  %sh0 = lshr i32 %x, 5
+  %r = or i32 %sh0, 42
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/shift-logic.ll:251]: It looks like the or gets optimized away entirely with this constant. You'll have to increase it so that the lshr by 7 does not make it zero.

File: llvm/test/Transforms/InstCombine/shift-logic.ll [hunk 14]
```diff
@@ -240,6 +240,35 @@ define i32 @lshr_or_extra_use(i32 %x, i32 %y, ptr %p) {
   ret i32 %sh1
 }
 
+define i32 @lshr_or_extra_use_multiuse(i32 %x, ptr %p) {
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/shift-logic.ll:243]:
```suggestion
define i32 @lshr_or_extra_use_shift(i32 %x, ptr %p) {
```
Or something. extra_use and multiuse are the same thing, so using both makes no sense.

**Comment by AtariDreams** [comment]
@nikic Done!

File: llvm/test/Transforms/InstCombine/shift-logic.ll [hunk 15]
```diff
@@ -240,6 +240,36 @@ define i32 @lshr_or_extra_use(i32 %x, i32 %y, ptr %p) {
...
+; CHECK-NEXT:    [[R:%.*]] = lshr i32 [[X]], 12
+; CHECK-NEXT:    [[SH1:%.*]] = or i32 [[R]], 192
+; CHECK-NEXT:    ret i32 [[SH1]]
+;
+  %sh0 = lshr i32 %x, 5
+  %r = or i32 %sh0, 24601
+  store i32 %sh0, ptr %p
+  %sh1 = lshr i32 %r, 7
+  ret i32 %sh1
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/shift-logic.ll:256]: It looks like this transform already happens without your patch: https://llvm.godbolt.org/z/6rbMWq65f

Is there some other fold that handles this pattern already?

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/shift-logic.ll:256]: The other tests you asked me to remove did in fact change, but not this one.

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/shift-logic.ll:256]: Can I please just have the other tests that do change?

**Comment by AtariDreams** [comment]
@nikic Can I use the other tests that I removed but is affected by the transform then?

**Comment by nikic** [comment]
You need the simplest possible test (no vectors...) that does not fold before your changes and folds after your changes. Please find this test on your own.

**Comment by AtariDreams** [comment]
> You need the simplest possible test (no vectors...) that does not fold before your changes and folds after your changes. Please find this test on your own.

Did. @nikic Is this good?

File: llvm/test/Transforms/InstCombine/shift-logic.ll [hunk 16]
```diff
@@ -433,9 +433,9 @@ define i8 @shl_sub(i8 %x, i8 %y) {
 define i8 @shl_sub_multiuse(i8 %x) {
 ; CHECK-LABEL: @shl_sub_multiuse(
 ; CHECK-NEXT:    [[SH0:%.*]] = shl i8 [[X:%.*]], 3
-; CHECK-NEXT:    [[R:%.*]] = sub i8 [[SH0]], 42 ; constant operand on the 'sub'
```
Comment (jrtc27) [inline llvm/test/Transforms/InstCombine/shift-logic.ll]: Uh, did update_llc_test_checks.py really add that comment?...

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/shift-logic.ll]: It removed the comment, not added it @nikic

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/shift-logic.ll]: It added the comment as it was in the original comment, but when it was transformed, llc removed it.

Comment (jrtc27) [reply llvm/test/Transforms/InstCombine/shift-logic.ll]: I know, but why was it there in the first place. The previous commit added it, but that suggests the first commit is wrong.

And please do not resolve conversations off the bat like that.

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/shift-logic.ll]: Because I put it there originally.

Comment (jrtc27) [reply llvm/test/Transforms/InstCombine/shift-logic.ll]: Well don't then?

And for the second time, stop resolving conversations like that.

**Comment by AtariDreams** [comment]
Alright @nikic Can we please merge this?

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp [hunk 17]
```diff
@@ -368,12 +368,11 @@ static Instruction *foldShiftOfShiftedBinOp(BinaryOperator &I,
 
   // Find a matching one-use shift by constant. The fold is not valid if the sum
```
Comment (topperc) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:369]: The "one-use" part of this comment is no longer completely accurate.

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp]: Fixed!

**Comment by AtariDreams** [comment]
@nikic Done!

**Review by nikic — APPROVED** [review]
LGTM

**Comment by LLVMCoCCommittee** [comment]
@AtariDreams We have reached out to request your feedback via the email attached to your GitHub account. Please respond to conduct@llvm.org
